tools/bin/check_images_exist.sh

```diff
@@ -1,78 +1,133 @@
+this_file_directory=$(dirname $0)
+relative_path_to_defaults=$this_file_directory/../shell_defaults
+
+# if a file exists there, source it. otherwise complain
+if test -f $relative_path_to_defaults; then
+  # source and '.' are the same program
+  source $relative_path_to_defaults
+else
+  echo -e "\033[31m\nFAILED TO SOURCE TEST RUNNING OPTIONS.\033[39m"
+  echo -e "\033[31mTried $relative_path_to_defaults\033[39m"
+  exit 1
+fi
+
+set +o xtrace  # +x easier human reading here
 
 . tools/lib/lib.sh
 
+
 function docker_tag_exists() {
   # Added check for images pushed to github container registry
   if [[ $1 == ghcr* ]]
   then
     TOKEN_URL=https://ghcr.io/token\?scope\="repository:$1:pull"
-    token=$(curl $TOKEN_URL | jq -r '.token')
+    token=$(curl $TOKEN_URL | jq -r '.token' > /dev/null)
     URL=https://ghcr.io/v2/$1/manifests/$2
-    printf "\tURL: %s\n" "$URL"
-    curl --silent -H "Authorization: Bearer $token" -f -lSL "$URL" > /dev/null
+    echo -e "$blue_text""\n\n\tURL: $URL""$default_text"
+    curl -H "Authorization: Bearer $token" --location --silent --show-error --dump-header header.txt "$URL" > /dev/null
+    curl_success=$?
   else
     URL=https://hub.docker.com/v2/repositories/"$1"/tags/"$2"
-    printf "\tURL: %s\n" "$URL"
-    curl --silent -f -lSL "$URL" > /dev/null
+    echo -e "$blue_text""\n\n\tURL: $URL""$default_text"
+    curl --silent --show-error --location --dump-header header.txt "$URL" > /dev/null
+    curl_success=$?
+    # some bullshit to get the number out of a header that looks like this
+    # < content-length: 1039
+    # < x-ratelimit-limit: 180
+    # < x-ratelimit-reset: 1665683196
+    # < x-ratelimit-remaining: 180
+    docker_rate_limit_remaining=$(grep 'x-ratelimit-remaining: ' < header.txt | grep --only-matching --extended-regexp "\d+")
+    # too noisy when set to < 1.  Dockerhub starts complaining somewhere around 10
+    if test "$docker_rate_limit_remaining" -lt 20; then
+      echo -e "$red_text""We are close to a sensitive dockerhub rate limit!""$default_text"
+      echo -e "$red_text""SLEEPING 60s sad times""$default_text"
+      sleep 60
+      docker_tag_exists $1 $2
+    elif test $docker_rate_limit_remaining -lt 50; then
+      echo -e "$red_text""Rate limit reported as $docker_rate_limit_remaining""$default_text"
+    fi
+  fi
+  if test $curl_success -ne 0; then
+    echo -e "$red_text""Curl Said this didn't work.  Please investigate""$default_text"
+    exit 1
   fi
 }
 
 checkPlatformImages() {
-  echo "Checking platform images exist..."
-  docker-compose pull || exit 1
-  echo "Success! All platform images exist!"
+  echo -e "$blue_text""Checking platform images exist...""$default_text"
+  #Pull without printing progress information and send error stream because that where image names are
+  docker-compose pull --quiet 2> compose_output
+  docker_compose_success=$?
+  images_pulled_count=$(docker images | wc -l)
+  if test $images_pulled_count -eq 0; then
```

Comment:
is this possible?  Would the headers from the output of `docker images` already put the line account above 0?

Also I guess if some prior action already built/pulled a docker image this wouldn't really tell us whether the `docker-compose pull` did anything, though that's not too important i think since this is just a sanity check.

Reply:
Good note, I don't think I double checked this logic and I didn't remove 100% of my images locally which is what I assume happens on a new cloud run

Reply:
after checking, failure is one line, but adding `--quiet` removes the header and is more clearly what we want to know